File: frontend/src/topology/locationBoundaryAnchors.ts
import type { FlowRectangle } from './nodeFootprint';
import type { MapCableRouteWaypoint, MapLocationBoundaryAnchor } from './savedMapTypes';

export interface BoundaryFrame { locationId: string; bounds: FlowRectangle }
type Point = Pick<MapCableRouteWaypoint, 'x' | 'y'>;
const clamp = (value: number) => Math.max(0, Math.min(1, value));
const near = (a: number, b: number) => Math.abs(a - b) <= 1e-7;
const inside = (point: Point, frame: FlowRectangle) => frame.x < point.x && point.x < frame.x + frame.width && frame.y < point.y && point.y < frame.y + frame.height;
const interpolate = (a: Point, b: Point, t: number): Point => ({ x: a.x + (b.x - a.x) * t, y: a.y + (b.y - a.y) * t });

export const resolveBoundaryWaypoint = (waypoint: MapCableRouteWaypoint, frames: readonly BoundaryFrame[]): MapCableRouteWaypoint => {
  const frame = waypoint.anchor && frames.find((item) => item.locationId === waypoint.anchor?.location_id)?.bounds;
  if (!frame) return waypoint;
  const { edge, offset } = waypoint.anchor!;
  return { ...waypoint, x: edge === 'left' ? frame.x : edge === 'right' ? frame.x + frame.width : frame.x + frame.width * offset,
    y: edge === 'top' ? frame.y : edge === 'bottom' ? frame.y + frame.height : frame.y + frame.height * offset };
};

export const projectBoundaryWaypoint = (locationId: string, frame: FlowRectangle, pointer: Point): MapCableRouteWaypoint => {
  const candidates: { edge: MapLocationBoundaryAnchor['edge']; point: Point; offset: number }[] = [
    { edge: 'top', point: { x: Math.max(frame.x, Math.min(frame.x + frame.width, pointer.x)), y: frame.y }, offset: clamp((pointer.x - frame.x) / frame.width) },
    { edge: 'right', point: { x: frame.x + frame.width, y: Math.max(frame.y, Math.min(frame.y + frame.height, pointer.y)) }, offset: clamp((pointer.y - frame.y) / frame.height) },
    { edge: 'bottom', point: { x: Math.max(frame.x, Math.min(frame.x + frame.width, pointer.x)), y: frame.y + frame.height }, offset: clamp((pointer.x - frame.x) / frame.width) },
    { edge: 'left', point: { x: frame.x, y: Math.max(frame.y, Math.min(frame.y + frame.height, pointer.y)) }, offset: clamp((pointer.y - frame.y) / frame.height) },
  ];
  const chosen = candidates.reduce((best, candidate) => Math.hypot(candidate.point.x - pointer.x, candidate.point.y - pointer.y) < Math.hypot(best.point.x - pointer.x, best.point.y - pointer.y) ? candidate : best);
  return { ...chosen.point, anchor: { location_id: locationId, edge: chosen.edge, offset: chosen.offset } };
};

/** Angular candidates are intersections of neighboring segment rays with the frame perimeter. */
export const assistBoundaryWaypoint = (
  locationId: string, frame: FlowRectangle, pointer: Point, neighbors: readonly Point[],
  flowToScreenPosition: (point: Point) => Point,
): MapCableRouteWaypoint => {
  const distance = (point: Point) => Math.hypot(flowToScreenPosition(point).x - flowToScreenPosition(pointer).x, flowToScreenPosition(point).y - flowToScreenPosition(pointer).y);
  for (const [step, capturePx] of [[45, 12], [15, 5]] as const) {
    const candidates: MapCableRouteWaypoint[] = [];
    for (const neighbor of neighbors) for (let angle = 0; angle < 360; angle += step) {
      const radians = angle * Math.PI / 180, dx = Math.cos(radians), dy = Math.sin(radians);
      for (const x of [frame.x, frame.x + frame.width]) {
        if (Math.abs(dx) < 1e-9) continue;
        const t = (x - neighbor.x) / dx, y = neighbor.y + t * dy;
        if (t > 1e-7 && y >= frame.y - 1e-7 && y <= frame.y + frame.height + 1e-7) candidates.push(projectBoundaryWaypoint(locationId, frame, { x, y }));
      }
      for (const y of [frame.y, frame.y + frame.height]) {
        if (Math.abs(dy) < 1e-9) continue;
        const t = (y - neighbor.y) / dy, x = neighbor.x + t * dx;
        if (t > 1e-7 && x >= frame.x - 1e-7 && x <= frame.x + frame.width + 1e-7) candidates.push(projectBoundaryWaypoint(locationId, frame, { x, y }));
      }
    }
    const closest = candidates.sort((left, right) => distance(left) - distance(right))[0];
    if (closest && distance(closest) <= capturePx) return closest;
  }
  return projectBoundaryWaypoint(locationId, frame, pointer);
};

const onBoundary = (point: Point, frame: FlowRectangle) =>
  ((near(point.x, frame.x) || near(point.x, frame.x + frame.width)) && frame.y <= point.y && point.y <= frame.y + frame.height) ||
  ((near(point.y, frame.y) || near(point.y, frame.y + frame.height)) && frame.x <= point.x && point.x <= frame.x + frame.width);

const crossings = (a: Point, b: Point, frame: FlowRectangle): { t: number; point: Point }[] => {
  const dx = b.x - a.x, dy = b.y - a.y;
  const parameters = [frame.x, frame.x + frame.width].flatMap((x) => {
    const t = dx ? (x - a.x) / dx : -1, y = a.y + t * dy;
    return t > 0 && t < 1 && y >= frame.y && y <= frame.y + frame.height ? [t] : [];
  }).concat([frame.y, frame.y + frame.height].flatMap((y) => {
    const t = dy ? (y - a.y) / dy : -1, x = a.x + t * dx;
    return t > 0 && t < 1 && x >= frame.x && x <= frame.x + frame.width ? [t] : [];
  }));
  return parameters.sort((left, right) => left - right).filter((t, index, sorted) => index === 0 || Math.abs(t - sorted[index - 1]) > 1e-9).flatMap((t) => {
    const before = interpolate(a, b, Math.max(0, t - 1e-6));
    const after = interpolate(a, b, Math.min(1, t + 1e-6));
    return inside(before, frame) !== inside(after, frame) ? [{ t, point: interpolate(a, b, t) }] : [];
  });
};

/** Explicit save/group-move normalization; read-only rendering never mutates routes. */
export const normalizeLocationBoundaryAnchors = (
  source: Point, target: Point, waypoints: readonly MapCableRouteWaypoint[], frames: readonly BoundaryFrame[],
): MapCableRouteWaypoint[] => {
  const resolved = waypoints.map((point) => resolveBoundaryWaypoint(point, frames));
  const explicitLocations = new Set(resolved.flatMap((point) => point.anchor ? [point.anchor.location_id] : []));
  const points = [source, ...resolved, target];
  const result: MapCableRouteWaypoint[] = [];
  for (let index = 0; index < points.length - 1; index += 1) {
    const a = points[index], b = points[index + 1];
    const events = frames.filter((frame) => !explicitLocations.has(frame.locationId)).flatMap((frame) => crossings(a, b, frame.bounds).map((crossing) => ({ ...crossing, frame })));
    for (const event of events.sort((left, right) => left.t - right.t)) result.push(projectBoundaryWaypoint(event.frame.locationId, event.frame.bounds, event.point));
    if (index >= resolved.length) continue;
    const waypoint = resolved[index];
    if (waypoint.anchor) { result.push(waypoint); continue; }
    const following = points[index + 2];
    const frame = frames.find((item) => !explicitLocations.has(item.locationId) && onBoundary(waypoint, item.bounds) && inside(a, item.bounds) !== inside(following, item.bounds) && !onBoundary(a, item.bounds) && !onBoundary(following, item.bounds));
    result.push(frame ? projectBoundaryWaypoint(frame.locationId, frame.bounds, waypoint) : waypoint);
  }
  return result;
};

File: app/location_boundary_anchors.py
"""Presentation-only Location boundary anchors for saved Cable routes."""

from __future__ import annotations

import uuid

from app.errors import ValidationError


def _inside(point: dict, frame: dict) -> bool:
    return frame["x"] < point["x"] < frame["x"] + frame["width"] and frame["y"] < point["y"] < frame["y"] + frame["height"]


def _on_boundary(point: dict, frame: dict) -> bool:
    left, top = frame["x"], frame["y"]
    right, bottom = left + frame["width"], top + frame["height"]
    near = lambda a, b: abs(a - b) <= 1e-7
    return ((near(point["x"], left) or near(point["x"], right)) and top <= point["y"] <= bottom or
            (near(point["y"], top) or near(point["y"], bottom)) and left <= point["x"] <= right)


def _project(location_id: uuid.UUID, frame: dict, point: dict) -> dict:
    left, top = frame["x"], frame["y"]
    right, bottom = left + frame["width"], top + frame["height"]
    candidates = [
        ("top", max(left, min(right, point["x"])), top, (max(left, min(right, point["x"])) - left) / frame["width"]),
        ("right", right, max(top, min(bottom, point["y"])), (max(top, min(bottom, point["y"])) - top) / frame["height"]),
        ("bottom", max(left, min(right, point["x"])), bottom, (max(left, min(right, point["x"])) - left) / frame["width"]),
        ("left", left, max(top, min(bottom, point["y"])), (max(top, min(bottom, point["y"])) - top) / frame["height"]),
    ]
    edge, x, y, offset = min(candidates, key=lambda item: (item[1] - point["x"]) ** 2 + (item[2] - point["y"]) ** 2)
    return {"x": x, "y": y, "anchor": {"location_id": str(location_id), "edge": edge, "offset": offset}}


def _resolve(point: dict, location_id: uuid.UUID, frame: dict) -> dict:
    anchor = point.get("anchor")
    if not anchor or str(anchor["location_id"]) != str(location_id):
        return dict(point)
    edge, offset = anchor["edge"], anchor["offset"]
    return {**point,
            "x": frame["x"] if edge == "left" else frame["x"] + frame["width"] if edge == "right" else frame["x"] + frame["width"] * offset,
            "y": frame["y"] if edge == "top" else frame["y"] + frame["height"] if edge == "bottom" else frame["y"] + frame["height"] * offset}


def normalize_boundary_route(
    location_id: uuid.UUID, frame: dict, source: dict, waypoints: list[dict], target: dict,
) -> tuple[list[dict], int]:
    """Prefer this Location's explicit splitter; materialize only for legacy routes."""
    explicit = [index for index, point in enumerate(waypoints)
                if point.get("anchor", {}).get("location_id") == str(location_id)]
    if len(explicit) > 1:
        raise ValidationError("Boundary route cannot be split unambiguously", {})
    resolved = [_resolve(point, location_id, frame) for point in waypoints]
    if explicit:
        return resolved, explicit[0]
    points = [source, *resolved, target]
    result: list[dict] = []
    for index, (start, end) in enumerate(zip(points, points[1:])):
        dx, dy = end["x"] - start["x"], end["y"] - start["y"]
        parameters = []
        for x in (frame["x"], frame["x"] + frame["width"]):
            if dx:
                t = (x - start["x"]) / dx
                y = start["y"] + t * dy
                if 0 < t < 1 and frame["y"] <= y <= frame["y"] + frame["height"]: parameters.append(t)
        for y in (frame["y"], frame["y"] + frame["height"]):
            if dy:
                t = (y - start["y"]) / dy
                x = start["x"] + t * dx
                if 0 < t < 1 and frame["x"] <= x <= frame["x"] + frame["width"]: parameters.append(t)
        ordered = sorted(parameters)
        unique = [t for position, t in enumerate(ordered) if position == 0 or abs(t - ordered[position - 1]) > 1e-9]
        for t in unique:
            before = {"x": start["x"] + dx * max(0, t - 1e-6), "y": start["y"] + dy * max(0, t - 1e-6)}
            after = {"x": start["x"] + dx * min(1, t + 1e-6), "y": start["y"] + dy * min(1, t + 1e-6)}
            if _inside(before, frame) != _inside(after, frame):
                result.append(_project(location_id, frame, {"x": start["x"] + dx * t, "y": start["y"] + dy * t}))
        if index >= len(resolved):
            continue
        waypoint = resolved[index]
        if waypoint.get("anchor"):
            result.append(waypoint)
            continue
        following = points[index + 2]
        if (_on_boundary(waypoint, frame) and _inside(start, frame) != _inside(following, frame)
                and not _on_boundary(start, frame) and not _on_boundary(following, frame)):
            result.append(_project(location_id, frame, waypoint))
        else:
            result.append(waypoint)
    anchors = [index for index, point in enumerate(result) if point.get("anchor", {}).get("location_id") == str(location_id)]
    if not anchors:
        raise ValidationError("Boundary Cable route does not cross LocationFrame", {})
    if len(anchors) > 1:
        raise ValidationError("Boundary route cannot be split unambiguously", {})
    return result, anchors[0]
